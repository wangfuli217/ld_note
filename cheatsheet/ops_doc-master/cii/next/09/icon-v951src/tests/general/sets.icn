#SRC: JCON

# set test

procedure main()
   local x, y, z

   wset("empty", x := set())
   write(type(x))
   write(image(?x))		# should fail
   write(image(member(x)))	# should fail
   wset("insert", insert(x))
   write(image(?x))		# should write &null
   write(image(member(x)))	# should write &null
   write(image(member(x,3)))	# should write &null
   wset("insert", insert(x))
   wset("delete", delete(x))
   wset("delete", delete(x))
   write()

   wset("x", x := set([1,2,4]))
   wset("y", y := set([1,2,5]))
   wset("x ++ y", x ++ y)
   wset("y ++ x", y ++ x)
   wset("x -- y", x -- y)
   wset("y -- x", y -- x)
   wset("x ** y", x ** y)
   wset("y ** x", y ** x)
   write()

   wset("empty", x := set(&null))
   wset("+ 1", insert(x, 1,4,7))	# only inserts 1
   wset("+ 2", insert(x, 2))
   wset("+ c", insert(x, "c"))
   wset("- 3", delete(x, 3))		# deletes nothing
   wset("- 1", delete(x, 1, 2))		# only deletes 1
   wset("- 1", delete(x, 1))
   wset("+ 2", insert(x, 2))
   wset("+ 1", insert(x, 1))
   wset("+ 7.0", insert(x, 7.0))
   wset("+ 7.0", insert(x, 7.0))
   wset("+ 'cs'", insert(x, 'cs'))
   wset("+ 'cs'", insert(x, 'cs'))
   wset("x =", x)
   write()

   wset("3,a,4", y := set([3,"a",4]))
   wset("y ++ x", y ++ x)
   wset("y ** x", y ** x)
   wset("y -- x", y -- x)
   wset("x -- y", x -- y)
   write()

   every insert(z := set(), !y)
   wset("z from !y", z)

   write()
   x := set([3,1,4,1,5,9,2,6,5,3,5])
   y := copy(x)
   delete(x, 4)
   insert(x, 7)
   insert(y, 0)
   delete(y, 1)
   wset("x", x)
   wset("y", y)
end



#  dump a set, assuming it contains nothing other than:
#	&null, 0 - 9, 0.0 - 9.0, "", "a" - "e", '', 'cs'

procedure wset(label, S)
   local x

   writes(right(label, 10), " :", right(*S, 2), " :")
   every x := &null | (0 to 9) | 0.0+(0 to 9) | "" | !"abcde" | '' | 'cs' do
      writes(" ", image(member(S, x)))
   write()
   return
end
