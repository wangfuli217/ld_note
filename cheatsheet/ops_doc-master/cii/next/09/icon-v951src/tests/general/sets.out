     empty : 0 :
set
    insert : 1 : &null
&null
&null
    insert : 1 : &null
    delete : 0 :
    delete : 0 :

         x : 3 : 1 2 4
         y : 3 : 1 2 5
    x ++ y : 4 : 1 2 4 5
    y ++ x : 4 : 1 2 4 5
    x -- y : 1 : 4
    y -- x : 1 : 5
    x ** y : 2 : 1 2
    y ** x : 2 : 1 2

     empty : 0 :
       + 1 : 1 : 1
       + 2 : 2 : 1 2
       + c : 3 : 1 2 "c"
       - 3 : 3 : 1 2 "c"
       - 1 : 2 : 2 "c"
       - 1 : 2 : 2 "c"
       + 2 : 2 : 2 "c"
       + 1 : 3 : 1 2 "c"
     + 7.0 : 4 : 1 2 7.0 "c"
     + 7.0 : 4 : 1 2 7.0 "c"
    + 'cs' : 5 : 1 2 7.0 "c" 'cs'
    + 'cs' : 5 : 1 2 7.0 "c" 'cs'
       x = : 5 : 1 2 7.0 "c" 'cs'

     3,a,4 : 3 : 3 4 "a"
    y ++ x : 8 : 1 2 3 4 7.0 "a" "c" 'cs'
    y ** x : 0 :
    y -- x : 3 : 3 4 "a"
    x -- y : 5 : 1 2 7.0 "c" 'cs'

 z from !y : 3 : 3 4 "a"

         x : 7 : 1 2 3 5 6 7 9
         y : 7 : 0 2 3 4 5 6 9
